// authing-js-sdk/build/module/lib/management/ManagementClient.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { OrgManagementClient } from './OrgManagementClient';
import { GraphqlClient } from '../common/GraphqlClient';
import { ManagementTokenProvider } from './ManagementTokenProvider';
import { UserPoolManagementClient } from './UserpoolManagementClient';
import { UsersManagementClient } from './UsersManagementClient';
import { sendEmail } from '../graphqlapi';
import { FastHttpClient, HttpClient } from '../common/HttpClient';
import Axios from 'axios';
import { RolesManagementClient } from './RolesManagementClient';
import { PoliciesManagementClient } from './PoliciesManagementClient';
import { UdfManagementClient } from './UdfManagementClient';
import { GroupsManagementClient } from './GroupsManagementClient';
import { AclManagementClient } from './AclManagementClient';
import { WhitelistManagementClient } from './WhitelistManagementClient';
import { StatisticsManagementClient } from './StatisticsManagementClient';
import jwtDecode from 'jwt-decode';
import { encrypt } from '../utils';
import { PublicKeyManager } from '../common/PublicKeyManager';
import { UserActionManagementClient } from './UserActionManagementClient';
import { ApplicationsManagementClient } from './ApplicationsManagementClient';
import { MFAManagementClient } from './MFAManagementClient';
import { PrincipalManagementClient } from './PrincipalManagement';
import { TenantManagementClient } from './TenantManagementClient';
var DEFAULT_OPTIONS = {
    timeout: 10000,
    onError: function (code, message) {
        throw { code: code, message: message };
    },
    host: 'https://core.authing.cn',
    requestFrom: 'sdk',
    encryptFunction: encrypt,
    headers: {
        'userpool-id': 'x-authing-userpool-id',
        'app-id': 'x-authing-app-id',
        'tenant-id': 'x-authing-app-tenant-id',
        'request-from': 'x-authing-request-from',
        'sdk-version': 'x-authing-sdk-version',
        lang: 'x-authing-lang'
    },
    lang: 'zh-CN'
};
var ManagementClient = /** @class */ (function () {
    function ManagementClient(options) {
        Object.keys(options).forEach(function (i) { return !options[i] && delete options[i]; });
        this.options = Object.assign({}, DEFAULT_OPTIONS, options);
        if (!this.options.userPoolId && !this.options.appId)
            throw new Error('请提供 userPoolId 或者 appId!');
        var graphqlApiEndpointV2 = this.options.host + "/graphql/v2";
        if (!this.options.secret && !this.options.accessToken) {
            this.options.onError(1000, 'Init Management Client failed, must provide at least secret or accessToken !');
        }
        Axios.defaults.baseURL = this.options.host;
        this.graphqlClient = new (this.options.graphqlClient || GraphqlClient)(graphqlApiEndpointV2, this.options);
        this.tokenProvider = new ManagementTokenProvider(this.options, this.graphqlClient);
        this.httpClient = new (this.options.httpClient || HttpClient)(this.options, this.tokenProvider);
        this.fastHttpClient = new (this.options.fastHttpClient || FastHttpClient)(this.options, this.tokenProvider);
        this.publicKeyManager = new PublicKeyManager(this.options, this.httpClient);
        this.users = new UsersManagementClient(this.options, this.graphqlClient, this.httpClient, this.fastHttpClient, this.tokenProvider, this.publicKeyManager);
        this.userpool = new UserPoolManagementClient(this.options, this.httpClient, this.graphqlClient, this.tokenProvider);
        this.org = new OrgManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
        this.roles = new RolesManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
        this.policies = new PoliciesManagementClient(this.options, this.graphqlClient, this.tokenProvider);
        this.udf = new UdfManagementClient(this.options, this.graphqlClient, this.tokenProvider);
        this.acl = new AclManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
        this.whitelist = new WhitelistManagementClient(this.options, this.graphqlClient, this.tokenProvider);
        this.groups = new GroupsManagementClient(this.options, this.graphqlClient, this.tokenProvider);
        this.statistics = new StatisticsManagementClient(this.options, this.httpClient, this.tokenProvider);
        this.userAction = new UserActionManagementClient(this.options, this.httpClient, this.tokenProvider);
        this.applications = new ApplicationsManagementClient(this.options, this.httpClient, this.graphqlClient, this.tokenProvider);
        this.mfa = new MFAManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
        this.principal = new PrincipalManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
        this.tenant = new TenantManagementClient(this.options, this.graphqlClient, this.httpClient, this.tokenProvider);
    }
    /**
     * @description 发送邮件
     * @param email: 邮件
     * @param scene: 发送场景
     *
     */
    ManagementClient.prototype.sendEmail = function (email, scene) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, sendEmail(this.graphqlClient, this.tokenProvider, { email: email, scene: scene })];
                    case 1:
                        data = (_a.sent()).sendEmail;
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @description 检测登录状态
     *
     */
    ManagementClient.prototype.checkLoginStatus = function (token, options) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, fetchUserDetail, decoded, data, id, user;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = options || {};
                        _a = options.fetchUserDetail, fetchUserDetail = _a === void 0 ? false : _a;
                        if (!token)
                            return [2 /*return*/, null];
                        decoded = null;
                        try {
                            decoded = jwtDecode(token);
                        }
                        catch (error) {
                            return [2 /*return*/, null];
                        }
                        data = decoded.data;
                        if (!!fetchUserDetail) return [3 /*break*/, 1];
                        return [2 /*return*/, data];
                    case 1:
                        id = data.id;
                        return [4 /*yield*/, this.users.detail(id)];
                    case 2:
                        user = _b.sent();
                        return [2 /*return*/, user];
                }
            });
        });
    };
    /**
     * @name isPasswordValid
     * @name_zh 检测密码是否合法
     * @description 检测密码是否合法
     *
     * @param {string} password 密码
     *
     * @example
     *
     * const envList = await managementClient.isPasswordValid('xxxxxx')
     *
     * @returns {Promise<{
     *  valid: boolean
     *  message: string
     * }>} 返回检测结果
     * @memberof ManagementClient
     */
    ManagementClient.prototype.isPasswordValid = function (password) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.httpClient.request({
                            method: 'POST',
                            url: this.options.host + "/api/v2/password/check",
                            data: {
                                password: password
                            }
                        })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * @description 设置语言
     */
    ManagementClient.prototype.setLang = function (lang) {
        this.options.lang = lang;
    };
    /**
     * @description 执行 GraphQL 请求
     */
    ManagementClient.prototype.makeGraphqlRequest = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var query, variables, token, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = options.query, variables = options.variables;
                        return [4 /*yield*/, this.tokenProvider.getToken()];
                    case 1:
                        token = _a.sent();
                        return [4 /*yield*/, this.graphqlClient.request({
                                query: query,
                                variables: variables,
                                token: token
                            })];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * @description 执行 RESTful 请求
     */
    ManagementClient.prototype.makeRestRequest = function (config) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.httpClient.request(config)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    return ManagementClient;
}());
export { ManagementClient };
