// authing-js-sdk/build/umd/lib/graphqlapi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../types/graphql.v2"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.searchNodes = exports.authorizedTargets = exports.authorizeResource = exports.listAuthorizedResources = exports.listNodeByCodeAuthorizedResources = exports.listNodeByIdAuthorizedResources = exports.listGroupAuthorizedResources = exports.listRoleAuthorizedResources = exports.listUserAuthorizedResources = exports.refreshAccessToken = exports.getAccessToken = exports.unbindEmail = exports.getUserGroups = exports.removeUserFromGroup = exports.addUserToGroup = exports.deleteGroups = exports.updateGroup = exports.createGroup = exports.groupWithUsersWithCustomData = exports.groupWithUsers = exports.group = exports.getGroups = exports.checkLoginStatus = exports.disablePolicyAssignment = exports.enablePolicyAssignment = exports.removePolicyAssignments = exports.addPolicyAssignments = exports.policyAssignments = exports.deletePolicies = exports.deletePolicy = exports.updatePolicy = exports.policy = exports.createPolicy = exports.policies = exports.deleteRoles = exports.deleteRole = exports.udf = exports.removeUdf = exports.setUdf = exports.removeUdv = exports.setUdfValueBatch = exports.setUdvBatch = exports.setUdv = exports.udfValueBatch = exports.udv = exports.addWhiteList = exports.removeWhiteList = exports.getWhiteList = exports.allow = exports.setMainDepartment = exports.getUserDepartments = exports.getUserRoles = exports.userBatchWithCustomData = exports.userBatch = exports.unbindPhone = exports.bindEmail = exports.bindPhone = exports.updateEmail = exports.updatePhone = exports.updatePassword = exports.isUserExists = exports.refreshToken = exports.removeMembers = exports.moveNode = exports.nodeById = exports.deleteNode = exports.updateNode = exports.addNode = exports.updateRole = exports.roleWithUsersWithCustomData = exports.roleWithUsers = exports.role = exports.roles = exports.deleteUsers = exports.deleteUser = exports.createUser = exports.findUserWithCustomData = exports.findUser = exports.searchUserWithCustomData = exports.searchUser = exports.updateUser = exports.updateUserpool = exports.archivedUsers = exports.usersWithCustomData = exports.users = exports.resetPasswordByForceResetToken = exports.resetPasswordByFirstLoginToken = exports.resetPassword = exports.sendFirstLoginVerifyEmail = exports.sendEmail = exports.getMembersByCode = exports.getMembersById = exports.moveMembers = exports.addMember = exports.revokeRole = exports.assignRole = exports.addRole = exports.registerByPhoneCode = exports.registerByUsername = exports.loginBySubAccount = exports.loginByPhonePassword = exports.loginByPhoneCode = exports.loginByUsername = exports.loginByEmail = exports.registerByEmail = exports.checkPasswordStrength = exports.isDenied = exports.deleteOrg = exports.createOrg = exports.userWithCustomData = exports.user = exports.getUserPoolDetail = exports.orgs = exports.rootNode = exports.isRootNode = exports.org = exports.getChildrenNodes = exports.isAllowed = void 0;
    var graphql_v2_1 = require("../types/graphql.v2");
    var isAllowed = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.IsActionAllowedDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.isAllowed = isAllowed;
    var getChildrenNodes = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ChildrenNodesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.getChildrenNodes = getChildrenNodes;
    var org = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.OrgDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.org = org;
    var isRootNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.IsRootNodeDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.isRootNode = isRootNode;
    var rootNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RootNodeDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.rootNode = rootNode;
    var orgs = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.OrgsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.orgs = orgs;
    var getUserPoolDetail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UserpoolDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.getUserPoolDetail = getUserPoolDetail;
    var user = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.user = user;
    var userWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UserWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.userWithCustomData = userWithCustomData;
    var createOrg = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CreateOrgDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.createOrg = createOrg;
    var deleteOrg = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteOrgDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.deleteOrg = deleteOrg;
    var isDenied = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.IsActionDeniedDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.isDenied = isDenied;
    var checkPasswordStrength = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CheckPasswordStrengthDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.checkPasswordStrength = checkPasswordStrength;
    var registerByEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RegisterByEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.registerByEmail = registerByEmail;
    var loginByEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.LoginByEmailDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.loginByEmail = loginByEmail;
    var loginByUsername = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.LoginByUsernameDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.loginByUsername = loginByUsername;
    var loginByPhoneCode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.LoginByPhoneCodeDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.loginByPhoneCode = loginByPhoneCode;
    var loginByPhonePassword = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.LoginByPhonePasswordDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.loginByPhonePassword = loginByPhonePassword;
    var loginBySubAccount = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.LoginBySubAccountDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.loginBySubAccount = loginBySubAccount;
    var registerByUsername = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.RegisterByUsernameDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.registerByUsername = registerByUsername;
    var registerByPhoneCode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            query = graphql_v2_1.RegisterByPhoneCodeDocument;
            token = tokenProvider.getToken();
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    token: token,
                    variables: variables
                })];
        });
    }); };
    exports.registerByPhoneCode = registerByPhoneCode;
    var addRole = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CreateRoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.addRole = addRole;
    var assignRole = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AssignRoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.assignRole = assignRole;
    var revokeRole = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RevokeRoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.revokeRole = revokeRole;
    var addMember = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AddMemberDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.addMember = addMember;
    var moveMembers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.MoveMembersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.moveMembers = moveMembers;
    var getMembersById = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.NodeByIdWithMembersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.getMembersById = getMembersById;
    var getMembersByCode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.NodeByCodeWithMembersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.getMembersByCode = getMembersByCode;
    var sendEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SendEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.sendEmail = sendEmail;
    var sendFirstLoginVerifyEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SendFirstLoginVerifyEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [4 /*yield*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    }); };
    exports.sendFirstLoginVerifyEmail = sendFirstLoginVerifyEmail;
    var resetPassword = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ResetPasswordDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.resetPassword = resetPassword;
    var resetPasswordByFirstLoginToken = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ResetPasswordByFirstLoginTokenDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.resetPasswordByFirstLoginToken = resetPasswordByFirstLoginToken;
    var resetPasswordByForceResetToken = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ResetPasswordByForceResetTokenDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.resetPasswordByForceResetToken = resetPasswordByForceResetToken;
    var users = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UsersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.users = users;
    var usersWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UsersWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.usersWithCustomData = usersWithCustomData;
    var archivedUsers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ArchivedUsersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.archivedUsers = archivedUsers;
    var updateUserpool = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateUserpoolDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateUserpool = updateUserpool;
    var updateUser = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateUserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateUser = updateUser;
    var searchUser = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SearchUserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.searchUser = searchUser;
    var searchUserWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SearchUserWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.searchUserWithCustomData = searchUserWithCustomData;
    var findUser = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.FindUserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.findUser = findUser;
    var findUserWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.FindUserWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.findUserWithCustomData = findUserWithCustomData;
    var createUser = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CreateUserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.createUser = createUser;
    var deleteUser = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteUserDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteUser = deleteUser;
    var deleteUsers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteUsersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteUsers = deleteUsers;
    var roles = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RolesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.roles = roles;
    var role = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.role = role;
    var roleWithUsers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RoleWithUsersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.roleWithUsers = roleWithUsers;
    var roleWithUsersWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RoleWithUsersWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.roleWithUsersWithCustomData = roleWithUsersWithCustomData;
    var updateRole = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateRoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateRole = updateRole;
    var addNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AddNodeV2Document;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.addNode = addNode;
    var updateNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateNodeDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateNode = updateNode;
    var deleteNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteNodeDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteNode = deleteNode;
    var nodeById = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.NodeByIdDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.nodeById = nodeById;
    var moveNode = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.MoveNodeDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.moveNode = moveNode;
    var removeMembers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemoveMemberDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removeMembers = removeMembers;
    var refreshToken = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RefreshTokenDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.refreshToken = refreshToken;
    var isUserExists = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.IsUserExistsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.isUserExists = isUserExists;
    var updatePassword = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdatePasswordDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updatePassword = updatePassword;
    var updatePhone = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdatePhoneDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updatePhone = updatePhone;
    var updateEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateEmail = updateEmail;
    var bindPhone = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.BindPhoneDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.bindPhone = bindPhone;
    var bindEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.BindEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.bindEmail = bindEmail;
    var unbindPhone = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UnbindPhoneDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.unbindPhone = unbindPhone;
    var userBatch = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UserBatchDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.userBatch = userBatch;
    var userBatchWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UserBatchWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.userBatchWithCustomData = userBatchWithCustomData;
    var getUserRoles = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GetUserRolesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.getUserRoles = getUserRoles;
    var getUserDepartments = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GetUserDepartmentsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.getUserDepartments = getUserDepartments;
    var setMainDepartment = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SetMainDepartmentDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.setMainDepartment = setMainDepartment;
    var allow = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AllowDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.allow = allow;
    var getWhiteList = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.WhitelistDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.getWhiteList = getWhiteList;
    var removeWhiteList = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemoveWhitelistDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removeWhiteList = removeWhiteList;
    var addWhiteList = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AddWhitelistDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.addWhiteList = addWhiteList;
    var udv = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UdvDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.udv = udv;
    var udfValueBatch = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UdfValueBatchDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.udfValueBatch = udfValueBatch;
    var setUdv = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SetUdvDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.setUdv = setUdv;
    var setUdvBatch = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SetUdvBatchDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.setUdvBatch = setUdvBatch;
    var setUdfValueBatch = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SetUdfValueBatchDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.setUdfValueBatch = setUdfValueBatch;
    var removeUdv = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemoveUdvDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removeUdv = removeUdv;
    var setUdf = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SetUdfDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.setUdf = setUdf;
    var removeUdf = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemoveUdfDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removeUdf = removeUdf;
    var udf = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UdfDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.udf = udf;
    var deleteRole = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteRoleDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteRole = deleteRole;
    var deleteRoles = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteRolesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteRoles = deleteRoles;
    var policies = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.PoliciesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.policies = policies;
    var createPolicy = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CreatePolicyDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.createPolicy = createPolicy;
    var policy = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.PolicyDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.policy = policy;
    var updatePolicy = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdatePolicyDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updatePolicy = updatePolicy;
    var deletePolicy = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeletePolicyDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deletePolicy = deletePolicy;
    var deletePolicies = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeletePoliciesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deletePolicies = deletePolicies;
    var policyAssignments = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.PolicyAssignmentsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.policyAssignments = policyAssignments;
    var addPolicyAssignments = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AddPolicyAssignmentsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.addPolicyAssignments = addPolicyAssignments;
    var removePolicyAssignments = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemovePolicyAssignmentsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removePolicyAssignments = removePolicyAssignments;
    var enablePolicyAssignment = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.EnablePolicyAssignmentDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.enablePolicyAssignment = enablePolicyAssignment;
    var disablePolicyAssignment = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DisbalePolicyAssignmentDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.disablePolicyAssignment = disablePolicyAssignment;
    var checkLoginStatus = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CheckLoginStatusDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.checkLoginStatus = checkLoginStatus;
    var getGroups = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GroupsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.getGroups = getGroups;
    var group = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GroupDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.group = group;
    var groupWithUsers = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GroupWithUsersDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.groupWithUsers = groupWithUsers;
    var groupWithUsersWithCustomData = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GroupWithUsersWithCustomDataDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.groupWithUsersWithCustomData = groupWithUsersWithCustomData;
    var createGroup = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.CreateGroupDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.createGroup = createGroup;
    var updateGroup = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UpdateGroupDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.updateGroup = updateGroup;
    var deleteGroups = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.DeleteGroupsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.deleteGroups = deleteGroups;
    var addUserToGroup = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AddUserToGroupDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.addUserToGroup = addUserToGroup;
    var removeUserFromGroup = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.RemoveUserFromGroupDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.removeUserFromGroup = removeUserFromGroup;
    var getUserGroups = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.GetUserGroupsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.getUserGroups = getUserGroups;
    var unbindEmail = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.UnbindEmailDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.unbindEmail = unbindEmail;
    var getAccessToken = function (garpqhlClient, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query;
        return __generator(this, function (_a) {
            query = graphql_v2_1.AccessTokenDocument;
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    variables: variables
                })];
        });
    }); };
    exports.getAccessToken = getAccessToken;
    var refreshAccessToken = function (garpqhlClient, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query;
        return __generator(this, function (_a) {
            query = graphql_v2_1.RefreshTokenDocument;
            return [2 /*return*/, garpqhlClient.request({
                    query: query,
                    variables: variables
                })];
        });
    }); };
    exports.refreshAccessToken = refreshAccessToken;
    var listUserAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ListUserAuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listUserAuthorizedResources = listUserAuthorizedResources;
    var listRoleAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ListRoleAuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listRoleAuthorizedResources = listRoleAuthorizedResources;
    var listGroupAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ListGroupAuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listGroupAuthorizedResources = listGroupAuthorizedResources;
    var listNodeByIdAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ListNodeByIdAuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listNodeByIdAuthorizedResources = listNodeByIdAuthorizedResources;
    var listNodeByCodeAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.ListNodeByCodeAuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listNodeByCodeAuthorizedResources = listNodeByCodeAuthorizedResources;
    var listAuthorizedResources = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AuthorizedResourcesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.listAuthorizedResources = listAuthorizedResources;
    var authorizeResource = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AuthorizeResourceDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.authorizeResource = authorizeResource;
    var authorizedTargets = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.AuthorizedTargetsDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.authorizedTargets = authorizedTargets;
    var searchNodes = function (garpqhlClient, tokenProvider, variables) { return __awaiter(void 0, void 0, void 0, function () {
        var query, token;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = graphql_v2_1.SearchNodesDocument;
                    return [4 /*yield*/, tokenProvider.getToken()];
                case 1:
                    token = _a.sent();
                    return [2 /*return*/, garpqhlClient.request({
                            query: query,
                            token: token,
                            variables: variables
                        })];
            }
        });
    }); };
    exports.searchNodes = searchNodes;
});
