// authing-js-sdk/build/main/lib/management/TenantManagementClient.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ManagementClient_1 = require("./ManagementClient");
var testing_helper_1 = require("../testing-helper");
var ava_1 = __importDefault(require("ava"));
var __1 = require("../..");
var managementClient = new ManagementClient_1.ManagementClient(testing_helper_1.getOptionsFromEnv());
ava_1.default('list', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenants;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.list({ page: 1, limit: 10 })];
            case 1:
                tenants = _a.sent();
                console.log(tenants);
                t.assert(tenants);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('details', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.details('619b07312d6b99e1af7d8e4e')];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('create', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.create({
                    name: "搜索",
                    appIds: "619b64e4ccc0467dcba00920",
                })];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('update', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.update('619b64fd2cfccd07a8296839', {
                    name: "聚合搜索"
                })];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('delete', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.delete('619b64fd2cfccd07a8296839')];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('config', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.config('619b64fd2cfccd07a8296839', { css: ".btnId {\n text-color: #FF00EE}" })];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('members', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.members('619b64fd2cfccd07a8296839', { page: 1, limit: 10 })];
            case 1:
                tenant = _a.sent();
                console.log(JSON.stringify(tenant));
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('addMembers', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.addMembers('619b64fd2cfccd07a8296839', ['619b07ab229e3bfa98e94ee2'])];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('removeMembers', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var tenant;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.removeMembers('619b64fd2cfccd07a8296839', '619b07ab229e3bfa98e94ee2')];
            case 1:
                tenant = _a.sent();
                console.log(tenant);
                t.assert(tenant);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('listExtIdp', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.listExtIdp('619b64fd2cfccd07a8296839')];
            case 1:
                result = _a.sent();
                console.log(JSON.stringify(result));
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('extIdpDetail', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.extIdpDetail('619b33a00412723ba777eabf')];
            case 1:
                result = _a.sent();
                console.log(JSON.stringify(result));
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('createExtIdp', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.createExtIdp({
                    tenantId: '619b64fd2cfccd07a8296839',
                    name: '微信身份源',
                    type: 'wechat',
                    connections: [{
                            type: 'wechat:pc',
                            identifier: 'wechatc1',
                            displayName: '微信身份源连接',
                            fields: { 'clientSecret': 'd1cuu12MdgGKrcItRyD6TeJLNqoWjRW0', 'clientID': 'cli_a1118cb96bf95013', 'displayName': '飞书身份源连接' },
                            userMatchFields: ['ss']
                        }]
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('updateExtIdp', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.updateExtIdp('619c84ce946e9c1913247af1', { 'name': '飞书身份源' })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('deleteExtIdp', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.deleteExtIdp('619c84ce946e9c1913247af1')];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('createExtIdpConnection', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.createExtIdpConnection({
                    extIdpId: '619c917f534a3b8ad988a209',
                    type: 'wechatmp-qrcode',
                    identifier: 'wechatc3',
                    displayName: '微信身份源连接1',
                    fields: { 'clientSecret': 'd1cuu12KrcItRyD6T', 'clientID': 'cli_a196bf9013', 'displayName': '飞书身份源连接1' },
                    userMatchFields: ['ss']
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('updateExtIdpConnection', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.updateExtIdpConnection('619c9490d7b1cec02bf982f6', {
                    displayName: '微信身份源连接2',
                    fields: { 'clientSecret': 'd1cuu12KrcItRyD6T', 'clientID': 'cli_a196bf9013', 'displayName': '飞书身份源连接1' },
                    userMatchFields: ['ss']
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('deleteExtIdpConnection', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.deleteExtIdpConnection('619cc337075fdb26f5fdbfa2')];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('checkExtIdpConnectionIdentifierUnique', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.checkExtIdpConnectionIdentifierUnique('wechatc4')];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('changeExtIdpConnectionState', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.changeExtIdpConnectionState('619cc337075fdb26f5fdbfa2', {
                    tenantId: '619b64fd2cfccd07a8296839',
                    enabled: true
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('batchChangeExtIdpConnectionState', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.batchChangeExtIdpConnectionState('619c917f534a3b8ad988a209', {
                    tenantId: '619b64fd2cfccd07a8296839',
                    enabled: true
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('setTanentAdmin', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.setTanentAdmin('619c917f534a3b8ad988a209', {
                    userIds: ['11', '22']
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('updateTenantMember', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.updateTenantMember('619c917f534a3b8ad988a209', '111', true)];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
ava_1.default('batchInsertResource', function (t) { return __awaiter(void 0, void 0, void 0, function () {
    var result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, managementClient.tenant.batchInsertResource({
                    bulk: [
                        {
                            code: '6666',
                            type: __1.ResourceType.Data,
                            description: "666",
                            actions: [
                                {
                                    name: "action666",
                                    description: "description666",
                                }
                            ],
                            apiIdentifier: '666',
                            namespace: 'default',
                        },
                        {
                            code: '777',
                            type: __1.ResourceType.Data,
                            description: "777",
                            actions: [
                                {
                                    name: "action777",
                                    description: "description777",
                                }
                            ],
                            apiIdentifier: '777',
                            namespace: 'default',
                        }
                    ]
                })];
            case 1:
                result = _a.sent();
                console.log(result);
                t.assert(result);
                return [2 /*return*/];
        }
    });
}); });
